plat/kvm/x86/multiboot2.c

```diff
@@ -0,0 +1,188 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <uk/essentials.h>
+#include <uk/arch/limits.h>
+#include <uk/arch/types.h>
+#include <uk/arch/paging.h>
+#include <uk/plat/bootstrap.h>
+#include <uk/plat/common/bootinfo.h>
+#include <uk/plat/common/lcpu.h>
+#include <uk/plat/common/memory.h>
+#include <uk/plat/common/sections.h>
+#include <uk/reloc.h>
+#include <kvm-x86/multiboot2.h>
+
+#include <errno.h>
+#include <string.h>
+
+#define multiboot2_crash(msg, rc)	ukplat_crash()
+
+#define MULTIBOOT_TAG_ALIGN_UP(x)	ALIGN_UP((x), MULTIBOOT_TAG_ALIGN)
+
+void _ukplat_entry(struct lcpu *lcpu, struct ukplat_bootinfo *bi);
+
+static inline void mrd_insert(struct ukplat_bootinfo *bi,
+			      const struct ukplat_memregion_desc *mrd)
+{
+	int rc;
+
+	if (unlikely(mrd->len == 0))
+		return;
+
+	rc = ukplat_memregion_list_insert(&bi->mrds, mrd);
+	if (unlikely(rc < 0))
+		multiboot2_crash("Cannot insert bootinfo memory region", rc);
+}
+
+void handle_mmap_tag(struct ukplat_memregion_desc mrd,
```

Comment:
What is the `mrd` argument for? I think you may as well declare it on this function's frame.